pageserver/src/tenant/gc_block.rs

```diff
@@ -42,6 +48,25 @@ impl GcBlock {
         }
     }
 
+    /// Blocks GC until `duration` has elapsed.
+    ///
+    /// We do this as the leases mapping are not persisted to disk. By delaying GC by default
+    /// length, we guarantee that all the leases we granted before will have a chance to renew
+    /// when we run GC for the first time after restart / transition from AttachedMulti to AttachedSingle.
+    pub(super) async fn block_for(&self, duration: Duration, cancel: &CancellationToken) {
+        {
+            let mut g = self.reasons.lock().unwrap();
+            g.tenant_post_attached_single_wait = true;
+        }
+
+        let _ = tasks::delay_by_duration(duration, cancel).await;
+
+        {
+            let mut g = self.reasons.lock().unwrap();
+            g.tenant_post_attached_single_wait = false;
+        }
+    }
```

Comment:
this is problematic still because we get into all kind of async cancellation problems:

dropping the future while it's doing the timed sleep would for example leave the this forever stuck in `tenant_post_attached_single_wait`.

we don't need this sleep at all nor the spawned task, we could just write down the deadline (`std::time::Instant::now() + lsn_lease_length`) before which we are blocked.

so I'd:
- get rid of this method
- replace it it with some simple per-use method for blocking gc until deadline because of this state transition (the thing which we are having hard time naming)
    - per-use method as in not generic "block_for" whatever

this way we don't need to worry about the tasks::delay_by_duration (still not needed IMHO), nor spawn a task.

Reply:
I did this in 7d8bbddbb990ad34452889b3dc1df09c03f084f3. If we are only blocked by the lsn lease deadline, does it make sense for us to sleep until the deadline has reached?

Reply:
We will skip current gc iteration if deadline not reached: 9ad70331f3c17dca0a6ba222a78ca6c44c525e1d